packages/@aws-cdk/core/lib/private/synthesis.ts

```diff
@@ -110,6 +111,36 @@ function prepareTree(root: IConstruct) {
   visit(root, 'post', construct => construct.onPrepare());
 }
 
+/**
+ * Find all stacks and add Metadata Resources to all of them
+ *
+ * There is no good generic place to do this. Can't do it in the constructor
+ * (because adding a child construct makes it impossible to set context on the
+ * node), and the generic prepare phase is deprecated.
+ *
+ * Only do this on [parent] stacks (not nested stacks), don't do this when
+ * disabled by the user.
+ *
+ * Also, only when running via the CLI. If we do it unconditionally,
+ * all unit tests everywhere are going to break massively. I've spent a day
+ * fixing our own, but downstream users would be affected just as badly.
+ *
+ * Stop at Assembly boundaries.
+ */
+function injectMetadataResources(root: IConstruct) {
+  visit(root, 'post', construct => {
+    if (!Stack.isStack(construct) || !construct._versionReportingEnabled) { return; }
+
+    // Because of https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/assert/lib/synth-utils.ts#L74
+    // synthesize() may be called more than once on a stack in unit tests, and the below would break
+    // if we execute it a second time. Guard against the constructs already existing.
+    const CDKMetadata = 'CDKMetadata';
+    if (construct.node.tryFindChild(CDKMetadata)) { return; }
+
+    new MetadataResource(construct, CDKMetadata);
```

Comment:
Remind me what's the rationale for not just adding this in the `Stack`'s constructor based on the prop?

Reply:
Can't set context anymore

Reply:
And why do we need to set context? My view is that the context is just used to determine the default for `analyticsReporting` when new stacks are created. From that point forward we shouldn't care about the context key, no?